segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,445 @@
+        long traceId = LoggerHelpers.traceEnter(log, "read", handle.getSegmentName(), offset, bufferOffset, length);
+
+        if (offset < 0 || bufferOffset < 0 || length < 0) {
+            throw new CompletionException(new ArrayIndexOutOfBoundsException());
+        }
+
+        try (InputStream reader = client.readObjectStream(config.getExts3Bucket(),
+                config.getExts3Root() + handle.getSegmentName(), Range.fromOffsetLength(offset, length))) {
+
+            if (reader == null) {
+                log.info("Object does not exist {} in bucket {} ", config.getExts3Root() + handle.getSegmentName(),
+                        config.getExts3Bucket());
+
+                throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+            }
+
+            int originalLength = length;
+
+            int bytesRead = 0;
+            while (length != 0) {
+                bytesRead = reader.read(buffer, bufferOffset, length);
+                length -= bytesRead;
+                bufferOffset += bytesRead;
+            }
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return originalLength;
+        }
+    }
+
+    private StreamSegmentInformation syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        S3ObjectMetadata result = client.getObjectMetadata(config.getExts3Bucket(),
+                config.getExts3Root() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
+
+        boolean canWrite = false;
+        canWrite = acls.getGrants().stream().filter((grant) -> {
+            return grant.getPermission().compareTo(Permission.WRITE) >= 0;
+        }).count() > 0;
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
+        } catch (Exception e) {
+            log.warn("Exception {} observed while getting segment info {}", e.getMessage(), streamSegmentName);
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (client.listObjects(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName)
```

Comment:
Doesn't `PutObjectRequest` throw if it's already there? Your check-and-create is not atomic